### PR TITLE
Add an option for renaming: rename by hash from the name

### LINKED ISSUE #12: Adding deterministic way to rename functions/variables
People who try to hide the fact of cheating often rename functions or variables names to their synonyms. For example, a very common case is to rename `it` to `iter` or `elem`, in the case of functions the function named `count` could be renamed to `cnt`, `compute` or `calculate`. Using such renamings may help the model to learn the similarity between commonly used tokens. In this regard, it is necessary to add renaming to a fixed name in `RenameEntitiesTransformation`. 
I guess we can somehow store all the synonymous groups and then during generating `newName` for variable or function in `RenameEntitiesVisitor` new name will be randomly sampled from the corresponding group

### PR DESCRIPTION
**Pull request by nbirillo**

### CODE REVIEW

File: src/transformations/RenameEntitiesTransformation.cpp [hunk 1]
```diff
@@ -11,13 +13,14 @@ RenameEntitiesVisitor::RenameEntitiesVisitor(Rewriter * rewriter, const bool ren
...
-                                             mt19937 * gen, const bool test):
+                                             mt19937 * gen, string hash_prefix, const bool with_hash,
+                                             const bool test):
         rewriter(rewriter), sm(rewriter->getSourceMgr()),
         rename_func(rename_func), rename_var(rename_var),
         token_len_generator(move(token_len_generator)),
         tokens_num_generator(move(tokens_num_generator)),
         char_generator(move(char_generator)),
-        gen(gen), test(test) {
+        gen(gen), hash_prefix(std::move(hash_prefix)), with_hash(with_hash), test(test)  {
```
Comment (maximzubkov) [inline src/transformations/RenameEntitiesTransformation.cpp]: Throughout the project, we tried to avoid direct calling to namespaces, that's why I think it will more preferable to rewrite it the following way:
```
hash_prefix(move(hash_prefix))
```

File: src/transformations/RenameEntitiesTransformation.cpp [hunk 2]
```diff
@@ -96,8 +99,12 @@ string RenameEntitiesVisitor::randomSnakeCaseName() {
     return newName;
 }
 
+string RenameEntitiesVisitor::hashName(string * name) {
+    return hash_prefix + "_" + std::to_string(std::hash<std::string>{}(*name));
```
Comment (maximzubkov) [inline src/transformations/RenameEntitiesTransformation.cpp]: The same comment about `std::` here

File: include/transformations/RenameEntitiesTransformation.h [hunk 3]
```diff
@@ -93,6 +100,11 @@ class RenameEntitiesTransformation : public ITransformation {
                                           const bool rename_var, const int seed,
                                           const int max_tokens, const int max_token_len,
                                           const bool test);
+
+    RenameEntitiesTransformation(const float p, const bool rename_func, const bool rename_var, const int seed,
```
Comment (maximzubkov) [inline include/transformations/RenameEntitiesTransformation.h]: I don't think it is necessary to create multiple constructors for `RenameEntitiesTransformation`. It will be better to add `const bool with_hash` argument to the existing constructor

File: include/transformations/RenameEntitiesTransformation.h [hunk 4]
```diff
@@ -30,14 +30,16 @@ std::discrete_distribution;
 /// RecursiveASTVisitor is a set of actions that are done
 /// when a certain node of AST is reached. RenameEntitiesVisitor in particular
 /// visits all entities (e.g functions or variables) and rename them to a
-/// random name
+/// random name (or name like prefix_(hash from old name)
```
Comment (maximzubkov) [inline include/transformations/RenameEntitiesTransformation.h]: As in the future, we are going to add another option -- renaming with semantically close names, it will be better to write it the following way:
```
/// visits all entities (e.g functions or variables) and rename them.
/// Currently supported renamings:
///    - random name
///    - hash taken from previous name with a sertein prefix
```

File: include/transformations/RenameEntitiesTransformation.h [hunk 5]
```diff
@@ -30,14 +30,16 @@ std::discrete_distribution;
...
 class RenameEntitiesVisitor : public RecursiveASTVisitor<RenameEntitiesVisitor> {
  public:
     explicit RenameEntitiesVisitor(Rewriter * rewriter, const bool rename_func, const bool rename_var,
                                    discrete_distribution<int> token_len_generator,
                                    discrete_distribution<int> tokens_num_generator,
                                    discrete_distribution<int> char_generator,
-                                   mt19937 * gen, const bool test = false);
+                                   mt19937 * gen, string  hash_prefix = "d",
```
Comment (maximzubkov) [inline include/transformations/RenameEntitiesTransformation.h]: Double space before `hash_prefix`, it is strange that `cpplint` missed it

File: include/transformations/RenameEntitiesTransformation.h [hunk 6]
```diff
@@ -69,8 +71,12 @@ class RenameEntitiesVisitor : public RecursiveASTVisitor<RenameEntitiesVisitor>
 
     const bool test = false;
 
+    const bool with_hash = false;
+    const string hash_prefix = "d";
+
     map<Decl *, string> decl2name;
     string randomSnakeCaseName();
+    string hashName(string *name);
```
Comment (maximzubkov) [inline include/transformations/RenameEntitiesTransformation.h]: I personally prefer adding spaces after `*`:
```
string hashName(string * name);
```

File: src/transformations/RenameEntitiesTransformation.cpp [hunk 7]
```diff
@@ -124,10 +132,13 @@ void RenameEntitiesASTConsumer::HandleTranslationUnit(ASTContext &ctx) {
 RenameEntitiesTransformation::RenameEntitiesTransformation(const float p, const bool rename_func,
                                                            const bool rename_var, const int seed,
                                                            const int max_tokens, const int max_token_len,
-                                                           const bool test):
+                                                           string  hash_prefix,
+                                                           const bool with_hash, const bool test):
         ITransformation(p, "rename entities"),
         rename_func(rename_func),
         rename_var(rename_var),
+        hash_prefix(std::move(hash_prefix)),
```
Comment (maximzubkov) [inline src/transformations/RenameEntitiesTransformation.cpp]: Same comment about `std::` here

File: src/transformations/RenameEntitiesTransformation.cpp [hunk 8]
```diff
@@ -96,8 +99,12 @@ string RenameEntitiesVisitor::randomSnakeCaseName() {
     return newName;
 }
 
+string RenameEntitiesVisitor::hashName(string * name) {
+    return hash_prefix + "_" + std::to_string(std::hash<std::string>{}(*name));
+}
+
 void RenameEntitiesVisitor::processVarDecl(Decl * decl, string * name) {
-    string randomName = test ? "test_" + *name : randomSnakeCaseName();
+    string randomName = test ? "test_" + *name : with_hash ? hashName(name) : randomSnakeCaseName();
```
Comment (maximzubkov) [inline src/transformations/RenameEntitiesTransformation.cpp]: I guess multiple single-line `if` statement is quite not easy to read. I suggest writing in the following manner:
```
string randomName;
if (test) {
   randomName = "test_" + *name;
} else {
   randomName = with_hash ? hashName(name) : randomSnakeCaseName();
}
```

File: src/transformations/RenameEntitiesTransformation.cpp [hunk 9]
```diff
@@ -166,6 +177,8 @@ ITransformation * RenameEntitiesTransformation::buildFromConfig(const YAML::Node
     const auto seed = config["seed"].as<int>();
     const auto max_tokens = config["max tokens"].as<int>();
     const auto max_token_len = config["max token len"].as<int>();
+    const auto hash_token = config["hash token"] != nullptr ? config["hash token"].as<string>() : "d";
```
Comment (maximzubkov) [inline src/transformations/RenameEntitiesTransformation.cpp]: I think it will be better to rename `hash token` to `has prefix`

File: src/transformations/RenameEntitiesTransformation.cpp [hunk 10]
```diff
@@ -166,6 +177,8 @@ ITransformation * RenameEntitiesTransformation::buildFromConfig(const YAML::Node
...
+    const auto with_hash = config["with hash"] != nullptr && config["with hash"].as<bool>();
```
Comment (maximzubkov) [inline src/transformations/RenameEntitiesTransformation.cpp]: Let's rewrite this as follows to make it more readble:
```
const auto with_hash =  (config["with hash"] != nullptr) ? config["with hash"].as<bool>() : false;
```
Also as a very nervous person, I would suggest using brackets both in this and in the previous line to avoid unexpected behavior due to operations priority

File: src/transformations/RenameEntitiesTransformation.cpp [hunk 11]
```diff
@@ -124,10 +132,13 @@ void RenameEntitiesASTConsumer::HandleTranslationUnit(ASTContext &ctx) {
 RenameEntitiesTransformation::RenameEntitiesTransformation(const float p, const bool rename_func,
                                                            const bool rename_var, const int seed,
                                                            const int max_tokens, const int max_token_len,
-                                                           const bool test):
+                                                           string  hash_prefix,
```
Comment (maximzubkov) [inline src/transformations/RenameEntitiesTransformation.cpp]: Another double space

File: include/transformations/RenameEntitiesTransformation.h (hunk 5 above)
Comment (maximzubkov) [inline include/transformations/RenameEntitiesTransformation.h]: Also, I think it is better to pass a pointer to `hash_prefix` to avoid copying. I remember your suggestion about moving from pointers to references, but let's use the pointer in the current pull request

**Review by maximzubkov — CHANGES_REQUESTED** [review]

**Review by maximzubkov — COMMENTED** [review]
Looks good! But as we discussed in slack, it will be better to separate the renaming functionality from AST rewritings

File: .github/workflows/ci.yml [hunk 12]
```diff
@@ -1,9 +1,6 @@
 name: CI
 
-on:
-  push:
-    branches: [ master ]
-  pull_request:
+on: [push, pull_request]
```
Comment (SpirinEgor) [inline .github/workflows/ci.yml]: Why did you change this? Running actions for every push would increase the number of spent minutes.

**Review by SpirinEgor — APPROVED** [review]
LGTM 👍
